examples/five_mem_node/main.rs

```diff
@@ -0,0 +1,346 @@
+        })
+        .count();
+
+    for th in handles {
+        th.join().unwrap();
+    }
+}
+
+struct Node {
+    // None if the raft is not initialized.
+    raft_group: Option<RawNode<MemStorage>>,
+    my_mailbox: Receiver<Message>,
+    mailboxes: HashMap<u64, Sender<Message>>,
+    // Key-value pairs after applied. `MemStorage` only contains raft logs,
+    // so we need an additional storage engine.
+    kv_pairs: HashMap<u16, String>,
+}
+
+impl Node {
+    // Create a raft leader only with itself in its configuration.
+    fn create_raft_leader(
+        id: u64,
+        my_mailbox: Receiver<Message>,
+        mailboxes: HashMap<u64, Sender<Message>>,
+    ) -> Self {
+        let mut cfg = example_config();
+        cfg.id = id;
+        cfg.peers = vec![id];
+        cfg.tag = format!("peer_{}", id);
+
+        let storage = MemStorage::new();
+        let raft_group = Some(RawNode::new(&cfg, storage, vec![]).unwrap());
+        Node {
+            raft_group,
+            my_mailbox,
+            mailboxes,
+            kv_pairs: Default::default(),
+        }
+    }
+
+    // Create a raft follower.
+    fn create_raft_follower(
+        id: u64,
+        my_mailbox: Receiver<Message>,
+        mailboxes: HashMap<u64, Sender<Message>>,
+    ) -> Self {
+        let mut cfg = example_config();
+        cfg.id = id;
+        cfg.tag = format!("peer_{}", id);
+
+        Node {
+            raft_group: None,
+            my_mailbox,
+            mailboxes,
+            kv_pairs: Default::default(),
+        }
+    }
+
+    // Initialize raft for followers.
+    fn initialize_raft_from_message(&mut self, msg: &Message) {
+        if !is_initial_msg(msg) {
+            return;
+        }
+        let mut cfg = example_config();
+        cfg.id = msg.get_to();
+        let storage = MemStorage::new();
+        self.raft_group = Some(RawNode::new(&cfg, storage, vec![]).unwrap());
+    }
+
+    // Step a raft message, initialize the raft if need.
+    fn step(&mut self, msg: Message) {
+        if self.raft_group.is_none() {
+            if is_initial_msg(&msg) {
+                self.initialize_raft_from_message(&msg);
+            } else {
+                return;
+            }
+        }
+        let raft_group = self.raft_group.as_mut().unwrap();
+        let _ = raft_group.step(msg);
+    }
+}
+
+fn on_ready(
+    raft_group: &mut RawNode<MemStorage>,
+    kv_pairs: &mut HashMap<u16, String>,
+    mailboxes: &HashMap<u64, Sender<Message>>,
+    proposals: &Mutex<VecDeque<Proposal>>,
+) {
+    if !raft_group.has_ready() {
+        return;
+    }
+    // Get the `Ready` with `RawNode::ready` interface.
+    let mut ready = raft_group.ready();
+
+    // Persistent raft logs. It's necessary because in `RawNode::advance` we stabilize
+    // raft logs to the latest position.
+    raft_group
+        .raft
+        .raft_log
+        .store
+        .wl()
+        .append(ready.entries())
+        .unwrap();
```

Comment:
I think we can make an issue for future work to make this example handle errors elegantly. :)

Reply:
Yes, you are right. We need to show users how to deal with different errors.